lisette/core.py

```diff
@@ -150,10 +150,20 @@ def fmt2hist(outp:str)->list:
             hist.append(tcr[1])
     return hist
 
+# %% ../nbs/00_core.ipynb
+def _skip_tools_cache(msgs, cache_idxs):
+    "Skip tool use blocks and tool results in and shift cache indices"
+    res = []
+    for idx in cache_idxs:
+        while msgs[idx].get('tool_calls', []) or msgs[idx]['role'] == 'tool': idx -= 1
```

Comment:
Some `cache_idxs` can end up being shifted back to the same idx. For example if last 4 messages are tool blocks / results and user sets `cahce_idxs=[-4,-3,-2,-1]` then it will be `cache_idxs=[-5,-5,-5,-5]`.

We can keep shifting them until all idxs are unique (if possible otherwise collision is unavoidable) but this might complicate the logic a bit?